apps/src/lib/util/speech.js

```diff
@@ -22,38 +20,149 @@ export function textToSpeech(
+  const forceHTML5 = false;
+  if (cachedBytes !== undefined && cachedBytes[gender] !== undefined) {
+    if (cachedBytes[gender].length > 0) {
+      Sounds.getSingleton().addPromiseToSpeechQueue(
+        cachedBytes[gender].slice(0),
+        {
+          volume: 1.0,
+          loop: false,
+          forceHTML5: forceHTML5,
+          allowHTML5Mobile: true,
+          onEnded: () => {
+            Sounds.getSingleton().onSpeechFinished();
+          }
+        }
+      );
+    } else {
+      Sounds.getSingleton().addPromiseToSpeechQueue(
+        undefined,
+        {
+          volume: 1.0,
+          loop: false,
+          forceHTML5: forceHTML5,
+          allowHTML5Mobile: true,
+          onEnded: () => {
+            Sounds.getSingleton().onSpeechFinished();
+          }
+        },
+        {
+          language: language,
+          text: text,
+          gender: gender,
+          profanityFoundCallback: profaneWords => {
+            asyncOutputWarning(
+              i18n.textToSpeechProfanity({
+                profanityCount: profaneWords.length,
+                profaneWords: profaneWords.join(', ')
+              })
+            );
+          }
+        }
+      );
+    }
+  } else {
+    let ssml = `<speak version="1.0" xmlns="https://www.w3.org/2001/10/synthesis" xml:lang="en-US"><voice name="${voice}">${text}</voice></speak>`;
+    const url = `https://${region}.tts.speech.microsoft.com/cognitiveservices/v1`;
+
+    var request = new XMLHttpRequest();
+    const resultPromise = new Promise(async function(resolve, reject) {
+      let profaneWords = await findProfanity(text, languageCode);
+      if (profaneWords && profaneWords.length > 0) {
+        if (!Array.isArray(profaneWords)) {
+          profaneWords = [profaneWords];
+        }
+        asyncOutputWarning(
+          i18n.textToSpeechProfanity({
+            profanityCount: profaneWords.length,
+            profaneWords: profaneWords.join(', ')
+          })
+        );
+        Sounds.getSingleton().registerTextBytes(
+          language,
+          text,
+          true,
+          profaneWords
+        );
+        resolve(PROFANITY_FOUND);
+      }
+      request.onreadystatechange = function() {
+        // Waiting for readyState to be 4 meaning that the request is done
+        if (request.readyState === 4) {
+          if (request.status >= 200 && request.status < 300) {
+            resolve(request.response);
+          } else {
+            reject({
+              status: request.status,
+              statusText: request.statusText
+            });
+          }
+        }
+      };
+
+      request.open('POST', url, true);
+      request.responseType = 'arraybuffer';
+      request.setRequestHeader('Authorization', 'Bearer ' + token);
+      request.setRequestHeader('Content-Type', 'application/ssml+xml');
+      request.setRequestHeader(
+        'X-Microsoft-OutputFormat',
+        'audio-16khz-32kbitrate-mono-mp3'
+      );
```

Comment:
@uponthesun @sureshc -- do either of you know if this request header will give us trouble in a specific environment (most likely production)? i'm not sure if/where we set the list of allowed request headers

Reply:
Is this request transmitted directly from the browser to the 3rd party web service API?

Reply:
We check for profanity before we make any other requests so I don't think they could skip the filtering

Reply:
> Is this request transmitted directly from the browser to the 3rd party web service API?

The request is transmitted from the browser to our servers (at `studio.code.org/profanity/find`), then to a 3rd party service